src/fpm_filesystem.F90

```diff
@@ -311,7 +345,107 @@ subroutine mkdir(dir)
     end if
 end subroutine mkdir
 
+#ifndef FPM_BOOTSTRAP
+!> Get file & directory names in directory `dir` using iso_c_binding.
+!!
+!!  - File/directory names return are relative to cwd, ie. preprended with `dir`
+!!  - Includes files starting with `.` except current directory and parent directory
+!!
+recursive subroutine list_files(dir, files, recurse, separator)
+    character(len=*), intent(in) :: dir
+    type(string_t), allocatable, intent(out) :: files(:)
+    logical, intent(in), optional :: recurse
+    character(len=1), optional :: separator
+
+    integer :: i
+    type(string_t), allocatable :: dir_files(:)
+    type(string_t), allocatable :: sub_dir_files(:)
+
+    type(c_ptr) :: dir_handle
+    type(c_ptr) :: dir_entry_c
+    character(len=:,kind=c_char), allocatable :: fortran_name
+    character(len=:), allocatable :: string_fortran
+    integer, parameter :: N_MAX = 256
+    type(string_t) :: files_tmp(N_MAX)
+    integer(kind=c_int) :: r
+    character(len=1) :: filesep
+
+    if (present(separator)) then
+        filesep = separator
+    else
+        select case (get_os_type())
+            case default
+                filesep = '/'
+            case (OS_WINDOWS)
+                filesep = '\'
+        end select
+    end if
```

Comment:
Can we use [`fpm_filesystem::join_path`](https://github.com/fortran-lang/fpm/blob/68937a4eae6a71b74edbf762c574cc2dc22bb2d6/src/fpm_filesystem.f90#L223) in this routine instead of repeating this logic?

Reply:
This code determines the file separator once and reuse the result later, `join_path` unconditionally calls `get_os_type` everytime it runs, can be inefficient when there are multiple files in the directory tree.

Reply:
Sorry to jump in with an unrelated question, would using `save` on the `filesep` in `join_path` help save some CPU cycles or is adding persistent state to a subroutine considered bad practice?

Reply:
> Sorry to jump in with an unrelated question, would using `save` on the `filesep` in `join_path` help save some CPU cycles or is adding persistent state to a subroutine considered bad practice?

I believe using `save` on the `filesep` in `join_path` help save some CPU cycles, yes. This technique would also work on the return variable of `get_os_type`, I think.

I don't know if these are good practices, though.

Reply:
I was going to suggest the same thing. I don't have an issue using `save` here. Perhaps adding the openmp ['threadprivate`](https://www.openmp.org/spec-html/5.0/openmpsu104.html) clause for the persistent variable(s) would be a good idea for future-proofing also.

Reply:
I would suggest that both `fpm_filesystem::join_path` and `get_os_type` both be modified to cache the things that don't change in the middle of execution (i.e. the OS type or filesep). Using the `save` attribute would be a reasonable approach.

The `save` attribute can *sometimes* be bad practice. You should definitely avoid using it in a way that causes a procedure to give different outputs given the same inputs. That wouldn't be the case here.

Reply:
The installer object is currently caching the output from `get_os_type` and reuses it whenever possible. This creates also the possibility to just overwrite the cached OS type in the testsuite to test the behaviour on Unix and Windows.

Reply:
> The `save` attribute can _sometimes_ be bad practice. You should definitely avoid using it in a way that causes a procedure to give different outputs given the same inputs. That wouldn't be the case here.

In most cases using the `save` attribute will introduce an if clause such as
```
integer, save :: os_type
logical, save :: os_unknown = .true.
!omp threadprivate(os_type, os_unknown)

if (os_unknown) then
  os_type = get_os_type()
  os_unknown = .false.
end if
```
which is an acceptable trade-off compared to going through the full logic of the OS type. 

In the case of the separator with only two possibilities `'/'` or `'\'` it would only eliminate one branch and one assignment. On second thought this is the type of micro-optimization I would really want to avoid until shown to be performance critical. Use of `save` in threaded code can be difficult to debug, as I've experienced with some ODE codes from Netlib.

Reply:
Thank you for the feedback. I've updated the PR with the suggestions. 

Regarding the file separator optimization, it looks like there's no consensus in here. I've modified both `get_os_type()` and `join_path`. If there is consensus in the opposite direction, I'll remove the optimization in `join_path` tomorrow.

Reply:
Thanks @brocolis - these changes look fine to me.  Ivan has a good point about the file separator optimisation being essentially redundant. The main optimisation needed is in `get_os_type()` which is now implemented, so I'll mark this as resolved.